src/audio/kpb.c

```diff
@@ -901,6 +901,130 @@ static int kpb_reset(struct comp_dev *dev)
+	size_t i;
+
+	buffer_stream_invalidate(source, size);
+
+	if (kpb->config.sampling_width == IPC4_DEPTH_16BIT) {
+		const ae_int16 *in_ptr = (const ae_int16 *)istream->r_ptr;
+		ae_int16x4 d16 = AE_ZERO16();
+		const size_t in_offset = kpb->num_of_in_channels_ * sizeof(ae_int16);
+		const size_t out_offset = kpb->mic_sel_cnt_ * sizeof(ae_int16);
+		const size_t samples_per_chan = size / (sizeof(uint16_t) * kpb->mic_sel_cnt_);
+		ae_int16 *buf = (ae_int16 *)(ostream->addr);
+		ae_int16 *buf_end = (ae_int16 *)(ostream->end_addr);
+		ae_int16 *out_ptr = (ae_int16 *)buf;
+
+		AE_SETCBEGIN0(buf);
+		AE_SETCEND0(buf_end);
+		for (ch = 0; ch < kpb->mic_sel_cnt_; ch++) {
+			const ae_int16 *input_data = (const ae_int16 *)(in_ptr) + kpb->offsets_[ch];
+
+			out_ptr = (ae_int16 *)ostream->w_ptr;
+			out_ptr += ch;
+			for (i = 0; i < samples_per_chan; i++) {
+				AE_L16_XP(d16, input_data, in_offset);
+				AE_S16_0_XC(d16, out_ptr, out_offset);
+			}
+		}
+	} else if (kpb->config.sampling_width == IPC4_DEPTH_32BIT) {
+		const ae_int32 *in_ptr = (const ae_int32 *)istream->r_ptr;
+		ae_int32x2 d32 = AE_ZERO32();
+		const size_t in_offset = kpb->num_of_in_channels_ * sizeof(ae_int32);
+		const size_t out_offset = kpb->mic_sel_cnt_ * sizeof(ae_int32);
+		const size_t samples_per_chan = size / (sizeof(uint32_t) * kpb->mic_sel_cnt_);
+		ae_int32 *buf = (ae_int32 *)(ostream->addr);
+		ae_int32 *buf_end = (ae_int32 *)(ostream->end_addr);
+		ae_int32 *out_ptr = (ae_int32 *)buf;
+
+		AE_SETCBEGIN0(buf);
+		AE_SETCEND0(buf_end);
+		for (ch = 0; ch < kpb->mic_sel_cnt_; ch++) {
+			const ae_int32 *input_data = (const ae_int32 *)(in_ptr) + kpb->offsets_[ch];
+
+			out_ptr = (ae_int32 *)ostream->w_ptr;
+			out_ptr += ch;
+			for (i = 0; i < samples_per_chan; i++) {
+				AE_L32_XP(d32, input_data, in_offset);
+				AE_S32_L_XC(d32, out_ptr, out_offset);
+			}
+		}
+	}
+}
+#else
+static void kpb_micselect_copy(struct comp_dev *dev, struct comp_buffer __sparse_cache *sink,
+			       struct comp_buffer __sparse_cache *source, size_t size,
+			       uint32_t in_channels)
+{
+	struct comp_data *kpb = comp_get_drvdata(dev);
+	struct audio_stream __sparse_cache *istream = &source->stream;
+	struct audio_stream __sparse_cache *ostream = &sink->stream;
+
+	buffer_stream_invalidate(source, size);
+	uint16_t out_channel = 0;
+	size_t out_samples;
+
+	if (kpb->config.sampling_width == IPC4_DEPTH_16BIT) {
+		const uint16_t *in_data;
+		uint16_t *out_data;
+		const uint32_t samples_per_chan = size / (sizeof(uint16_t) * kpb->mic_sel_cnt_);
+
+		for (uint32_t ch = 0; ch < in_channels; ch++) {
+			if (IS_BIT_SET(kpb->mic_sel_->mask_, ch)) {
+				out_samples = 0;
+				in_data = (uint16_t *)istream->r_ptr;
+				out_data = (uint16_t *)ostream->w_ptr;
+
+				for (size_t i = 0; i < samples_per_chan; i++) {
+					if (&out_data[out_samples * kpb->mic_sel_cnt_ + out_channel]
+							>= (uint16_t *)ostream->end_addr) {
+						out_data = (uint16_t *)ostream->addr;
+						out_samples = 0;
+					}
```

Comment:
I think you could optimise this too using our new strategy: pre-calculate how many samples fit until the end of the buffer, copy that, then wrap and copy the rest? You could make the copying loop into a small function to not duplicate the code